File: 20-ValidParentheses/20-ValidParentheses.cs
// Last updated: 17/08/2026, 17:34:42
1public class Solution {
2    private Dictionary<char, char> brackets;
3
4    public Solution() {
5        brackets = new Dictionary<char, char> {
6            {')', '('},
7            {']', '['},
8            {'}', '{'},
9        };
10    }
11
12    public bool IsValid(string s) {
13        var stack = new Stack<char>();
14
15        foreach (var character in s) {
16            if (brackets.ContainsKey(character)) {
17                if (stack.Count == 0 || stack.Pop() != brackets[character]) {
18                    return false;
19                }
20            }
21            else {
22                stack.Push(character);
23            }
24        }
25
26        return stack.Count == 0;
27    }
28}

File: 20-ValidParentheses/20-ValidParentheses.js
// Last updated: 17/08/2026, 17:19:29
1/**
2 * @param {string} s
3 * @return {boolean}
4 */
5var isValid = function(s) {
6    if (!s) {
7        return false;
8    }
9
10    const brackets = {
11        ")": "(",
12        "}": "{",
13        "]": "[",
14    };
15
16    const stack = [];
17
18    for (let char of s) {
19        if (brackets[char]) {
20            const popChar = stack.length && stack.pop();
21            
22            if (popChar !== brackets[char]) {
23                return false;
24            }
25        }
26        else {
27            stack.push(char);
28        }
29    }
30
31    return stack.length === 0;
32};